kv/memberlist/memberlist_client_test.go

```diff
@@ -1710,3 +1717,86 @@ func getKey(t *testing.T, msg []byte) string {
 	require.NoError(t, err)
 	return kvPair.Key
 }
+
+func TestRaceBetweenStoringNewValueForKeyAndUpdatingIt(t *testing.T) {
```

Comment:
Does this consistently catch the bug for you? It consistently does not for me, I tried running it a 100 times with your fix reverted.

Reply:
Interesting. Yes, I can reproduce reliably on every run.

Reply:
Never mind, I have to use the `-race` flag of course.

Reply:
```
GOROOT=/opt/homebrew/opt/go/libexec #gosetup
GOPATH=/Users/peter/go #gosetup
/opt/homebrew/opt/go/libexec/bin/go test -c -race -o /Users/peter/Library/Caches/JetBrains/GoLand2024.2/tmp/GoLand/___TestRaceBetweenStoringNewValueForKeyAndUpdatingIt_in_github_com_grafana_dskit_kv_memberlist.test github.com/grafana/dskit/kv/memberlist #gosetup
/opt/homebrew/opt/go/libexec/bin/go tool test2json -t /Users/peter/Library/Caches/JetBrains/GoLand2024.2/tmp/GoLand/___TestRaceBetweenStoringNewValueForKeyAndUpdatingIt_in_github_com_grafana_dskit_kv_memberlist.test -test.v=test2json -test.paniconexit0 -test.run ^\QTestRaceBetweenStoringNewValueForKeyAndUpdatingIt\E$
=== RUN   TestRaceBetweenStoringNewValueForKeyAndUpdatingIt
==================
WARNING: DATA RACE
Read at 0x00c0003ee7e0 by goroutine 51:
  runtime.mapdelete()
      /opt/homebrew/opt/go/libexec/src/runtime/map.go:741 +0x43c
  github.com/grafana/dskit/kv/memberlist.TestRaceBetweenStoringNewValueForKeyAndUpdatingIt.func2()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client_test.go:1756 +0x314

Previous write at 0x00c0003ee7e0 by goroutine 10:
  runtime.mapaccess2_faststr()
      /opt/homebrew/opt/go/libexec/src/runtime/map_faststr.go:117 +0x42c
  github.com/grafana/dskit/kv/memberlist.(*data).Merge()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client_test.go:61 +0x238
  github.com/grafana/dskit/kv/memberlist.computeNewValue()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client.go:1461 +0xc4
  github.com/grafana/dskit/kv/memberlist.(*KV).mergeValueForKey()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client.go:1404 +0x1c8
  github.com/grafana/dskit/kv/memberlist.(*KV).mergeBytesValueForKey()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client.go:1385 +0x140
  github.com/grafana/dskit/kv/memberlist.(*KV).MergeRemoteState()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client.go:1345 +0x30c
  github.com/grafana/dskit/kv/memberlist.TestRaceBetweenStoringNewValueForKeyAndUpdatingIt()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client_test.go:1777 +0x6a0
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x40

Goroutine 51 (running) created at:
  github.com/grafana/dskit/kv/memberlist.TestRaceBetweenStoringNewValueForKeyAndUpdatingIt()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client_test.go:1742 +0x58c
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x40

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x5e0
  testing.runTests.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2168 +0x80
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1690 +0x184
  testing.runTests()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2166 +0x6e0
  testing.(*M).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2034 +0xb74
  main.main()
      _testmain.go:105 +0x110
==================
==================
WARNING: DATA RACE
Read at 0x00c0003f9718 by goroutine 51:
  github.com/grafana/dskit/kv/memberlist.TestRaceBetweenStoringNewValueForKeyAndUpdatingIt.func2()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client_test.go:1756 +0x384

Previous write at 0x00c0003f9718 by goroutine 10:
  github.com/grafana/dskit/kv/memberlist.(*data).Merge()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client_test.go:61 +0x248
  github.com/grafana/dskit/kv/memberlist.computeNewValue()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client.go:1461 +0xc4
  github.com/grafana/dskit/kv/memberlist.(*KV).mergeValueForKey()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client.go:1404 +0x1c8
  github.com/grafana/dskit/kv/memberlist.(*KV).mergeBytesValueForKey()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client.go:1385 +0x140
  github.com/grafana/dskit/kv/memberlist.(*KV).MergeRemoteState()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client.go:1345 +0x30c
  github.com/grafana/dskit/kv/memberlist.TestRaceBetweenStoringNewValueForKeyAndUpdatingIt()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client_test.go:1777 +0x6a0
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x40

Goroutine 51 (running) created at:
  github.com/grafana/dskit/kv/memberlist.TestRaceBetweenStoringNewValueForKeyAndUpdatingIt()
      /Users/peter/Grafana/dskit/kv/memberlist/memberlist_client_test.go:1742 +0x58c
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1690 +0x184
  testing.(*T).Run.gowrap1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x40

Goroutine 10 (running) created at:
  testing.(*T).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1743 +0x5e0
  testing.runTests.func1()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2168 +0x80
  testing.tRunner()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:1690 +0x184
  testing.runTests()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2166 +0x6e0
  testing.(*M).Run()
      /opt/homebrew/opt/go/libexec/src/testing/testing.go:2034 +0xb74
  main.main()
      _testmain.go:105 +0x110
==================
    testing.go:1399: race detected during execution of test
--- FAIL: TestRaceBetweenStoringNewValueForKeyAndUpdatingIt (1.02s)

FAIL

Process finished with the exit code 1

```